lang/src/org/partiql/lang/eval/EvaluatingCompiler.kt

```diff
@@ -192,30 +191,26 @@ internal class EvaluatingCompiler(
             }
         }
 
-        override fun compute() = current?.exprValue() ?: valueFactory.nullValue
+        override fun compute() = current ?: valueFactory.nullValue
     }
 
-    private fun comparisonAccumulator(cmpFunc: (Number, Number) -> Boolean): (Number?, ExprValue) -> Number =
-        { curr, next ->
-            val nextNum = next.numberValue()
-            when (curr) {
-                null -> nextNum
-                else -> when {
-                    cmpFunc(nextNum, curr) -> nextNum
-                    else -> curr
-                }
+    private fun comparisonAccumulator(comparator: NaturalExprValueComparators): (ExprValue?, ExprValue) -> ExprValue =
+        { left, right ->
+            when {
+                left == null || comparator.compare(left, right) > 0 -> right
+                else -> left
             }
         }
 
     /** Dispatch table for built-in aggregate functions. */
     private val builtinAggregates: Map<Pair<String, PartiqlAst.SetQuantifier>, ExprAggregatorFactory> =
         run {
-            val countAccFunc: (Number?, ExprValue) -> Number = { curr, _ -> curr!! + 1L }
-            val sumAccFunc: (Number?, ExprValue) -> Number = { curr, next ->
-                curr?.let { it + next.numberValue() } ?: next.numberValue()
+            val countAccFunc: (ExprValue?, ExprValue) -> ExprValue = { accumulated, _ -> (accumulated!!.longValue() + 1L).exprValue() }
+            val sumAccFunc: (ExprValue?, ExprValue) -> ExprValue = { accumulated, nextItem ->
```

Comment:
nit: after playing around in the REPL/CLI a bit, I think the error messages could be improved for the `SUM` and `AVG` aggregation functions involving non-numeric types. Currently when giving a non-numeric type, we get something like the following error:

```
org.partiql.lang.eval.EvaluationException: Expected number: "abc"
```

Could consider providing a more descriptive error message (e.g. "SUM expects numeric types but was given a STRING 'abc'") in this PR or cutting a separate issue.